### PR TITLE
Update gather documentation to allow index.shape[k] <= input.shape[k] rather than ==.

### PR DESCRIPTION
**Pull request by gchanan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41887 Update gather documentation to allow index.shape[k] <= input.shape[k] rather than ==.**

Differential Revision: [D22680014](https://our.internmc.facebook.com/intern/diff/D22680014)

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

**Comment by gchanan** [comment]
This dropped the statement that `shape[dim] > 1`, need to check if that's accurate.

**Comment by gchanan** [comment]
Verification that dropping `index.shape[dim] >= 1` is okay:

```
>>> torch.gather(torch.randn(4, 2), dim=1, index=torch.zeros(4, 0))
tensor([], size=(4, 0))
>>> torch.gather(torch.randn(4, 0), dim=1, index=torch.zeros(4, 0))
tensor([], size=(4, 0))
>>> torch.gather(torch.randn(4, 2).cuda(), dim=1, index=torch.zeros(4, 0).cuda())
tensor([], device='cuda:0', size=(4, 0))
>>> torch.gather(torch.randn(4, 0).cuda(), dim=1, index=torch.zeros(4, 0).cuda())
tensor([], device='cuda:0', size=(4, 0))
```
